core/drivers/crypto/hisilicon/hisi_qm.c

```diff
@@ -143,49 +147,108 @@ static void qm_db(struct hisi_qm *qm, uint16_t qn, uint8_t cmd, uint16_t index,
 	io_write64(qm->io_base + QM_DOORBELL_SQ_CQ_BASE, doorbell);
 }
 
+static void qm_mb_write(struct hisi_qm *qm, struct qm_mailbox *mb)
+{
+	vaddr_t dst = qm->io_base + QM_MAILBOX_BASE;
+
+	write_64bit_pair(dst, mb->x[0], mb->x[1]);
+}
+
+static void qm_mb_read(struct hisi_qm *qm, struct qm_mailbox *mb)
+{
+	vaddr_t mb_base = qm->io_base + QM_MAILBOX_BASE;
+
+	write_64bit_pair((vaddr_t)mb, *(uint64_t *)mb_base,
```

Comment:
> The hardware requires that 128bits should be read from hardware at one time.It is implemented in kernel in that way using stp as picture above.My modification was wrong.Solved now.

This doesn't make sense. If you need an atomic 128-bit read, then you should do an `ldp` or `read_64bit_pair()` not a `stp` or `write_64bit_pair()`. The compiler might implement the two memory dereferences `*(uint64_t *)mb_base` and `*((uint64_t *)mb_base + 1)` with an `ldp` instruction, but that's not guaranteed. `mb` only resides on the stack when this function is called from `qm_wait_mb_ready()` and `qm_wait_mb_finish()`, so that struct should be accessible in a normal way.

By the way, I don't think it makes sense to submit untested code upstream. Not tested == not working, we've seen that before.

Reply:
The code has been tested.This function actually writes from mb_base to mb.When tested, I found a problem in qm_mb_write and solved it.

Reply:
Good that you tested it and could fix problems because of that, but you still haven't addressed my other comment:

> This doesn't make sense. If you need an atomic 128-bit read, then you should do an ldp or read_64bit_pair() not a stp or write_64bit_pair(). The compiler might implement the two memory dereferences *(uint64_t *)mb_base and *((uint64_t *)mb_base + 1) with an ldp instruction, but that's not guaranteed. mb only resides on the stack when this function is called from qm_wait_mb_ready() and qm_wait_mb_finish(), so that struct should be accessible in a normal way.

Reply:
mb_base and mb are both memory.I need use stp to write info to memory.I use  *(uint64_t *)mb_base and *((uint64_t *)mb_base + 1 to implement ldp，but stp is needed in final.

Reply:
The original code is implemented in this way. stp is needed. I only change the way to implement ldp.

Reply:
![image](https://github.com/OP-TEE/optee_os/assets/118411745/fd20f561-b14c-4b2f-919e-096bbb67a447)

Reply:
I can see how the `stp` could be needed when writing to the shared mailbox, but when reading from the shared mailbox I don't see why you need an `stp` instruction when saving the read value on the stack. I'd expect `ldp` to be used when reading from the shared mailbox if you intend to use 128-bit atomic memory accesses to the shared mailbox.

Reply:
solved